supervisor/api/backups.py

```diff
@@ -518,29 +519,28 @@ async def upload(self, request: web.Request):
             except vol.Invalid as ex:
                 raise APIError(humanize_error(filename, ex)) from None
 
-        with TemporaryDirectory(dir=tmp_path.as_posix()) as temp_dir:
-            tar_file = Path(temp_dir, "backup.tar")
+        temp_dir: TemporaryDirectory | None = None
+        backup_file_stream: IOBase | None = None
+
+        def open_backup_file() -> Path:
+            nonlocal temp_dir, backup_file_stream
```

Comment:
Similar to above, we could make the error handling part of `open_backup_file`. The advantage is that the "consumer" of the function doesn't need to care. As in, if the function returns, then the consumer is responsible for cleanup. If not, we can consider things to be clean.

It is a bit a style question, either case should technically be fine.

Reply:
Actually its not just style here. I originally wrote it that way because I also prefer to use returns. The reason I switched is because if the file open fails in this function then it raises and there is no return. This leaves us unable to clean up the temporary directory we created. The alternative seemed to be creating the temporary directory and opening the file in two separate executor tasks but I picked less clean code over worse performance as seemingly the lesser of two evils.

I suppose I could've used a return for `backup_file_stream` rather then a `nonlocal` like this. But I guess once I realized I had to use `nonlocal` for `temp_dir` I decided to not mix and match output methods.

Reply:
After discussion out-of-band, we agreed that the two should be technically equivalent. Since functions are inline and not meant to be used separately, making them taking care of cleanup themselfs is not really that big of an advantage. Having cleanup always in the main function might also be seen as improving overall readability. In any case, I am fine either way. So let's consider this closed.